diesel/src/mysql/connection/mod.rs

```diff
@@ -248,8 +248,10 @@ fn prepared_query<'a, T: QueryFragment<Mysql> + QueryId>(
     statement_cache: &'a mut StatementCache<Mysql, Statement>,
     raw_connection: &'a mut RawConnection,
 ) -> QueryResult<MaybeCached<'a, Statement>> {
-    let mut stmt = statement_cache
-        .cached_statement(source, &Mysql, &[], |sql, _| raw_connection.prepare(sql))?;
+    let mut stmt =
+        statement_cache.cached_statement(T::query_id(), source, &Mysql, &[], &mut |sql, _| {
```

Comment:
in order to avoid having to update all these places and guarantee consistency between source and maybe_type_id everywhere, we can leave the signature of the cached_statement function as it is, and immediately call an inner function that is not generic.
The cached_statement function will always be inlined, so the generated code will be the same as this, but we keep the previous (safer) interface and abstract the "please only monomorphize once" internals.